text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,100 @@
+
+- Start Date: 2022-08-12
+- RFC PR: (leave this empty)
+- FOLIO Issue: (leave this empty)
+
+# Cease Creating Kafka Topics For Each Tenant
+
+## Summary
+
+The purpose of this RFC is to lay out the reasoning for creating Kafka topics using a different scheme other than what is currently employed; creating the same topic for each tenant. The recommendation is to create single topic for each application concern. This will reduce the partition count needed for normal operations.
+
+## Motivation
+FOLIO should cease creating Kafka topics for each tenant installed. This has scaling & cost implications.
+
+- **Performance**: 
+Similar to a database not having a theoretical limit on the number of rows/tables, Kafka does not have a limit on partition count. A major difference is that a partition is "heavier" than a row/table. Partitions have to be rebalanced, replicated with corresponding open server files for each partition. Partitions have an ongoing administrative cost while online. Creating numerous partitions like they are database rows/tables will spell disaster for an under-provisioned Kafka installation. This is why cloud providers ensure to have limits on partition counts. Provisioning a Kafka installation to support such a high partition count will be wasteful for FOLIO hosting providers.
```

Comment:
The existence of a partition has an overhead. Having topics/partitions for each tenant means that the overhead burden is incurred while having minimal value if the topic is not used often. Example, a tenant does a data import load once, all topics are created. Tenant doesn’t do imports at all or not as often so topics are lingering have to be managed by kafka. Compared to one application topic with that will be well utilized because it is in use by multiple tenants. While having a 1000 tenants, it is more efficient for kafka to manage one topic with 1000 messages than 1000 topics with a message each.

Reply:
> The existence of a partition has an overhead. 

Sure, my question related to what that overhead is.

> Having topics/partitions for each tenant means that the overhead burden is incurred while having minimal value if the topic is not used often.

When you say the topic isn't used often, do you mean for a specific tenant or generally?

I'm not sure I understand how frequency of use is a factor in the performance concerns you refer to.

> Tenant doesn’t do imports at all or not as often so topics are lingering have to be managed by kafka.

This might be a system configuration / design rabbit hole. If a tenant does not do imports, it should be able to avoid having the topics specific to data import. That it cannot, might be an trade off of a separate design decision.

> While having a 1000 tenants, it is more efficient for kafka to manage one topic with 1000 messages than 1000 topics with a message each.

It might be more efficient from a storage perspective in Kafka. A trade off with that is that it potentially allows activity in one tenant to degrade the experience in another tenant, e.g. a large import in one tenant could hold up other imports in other tenants.

Reply:
>Sure, my question related to what that overhead is.

Ah, the overhead includes more objects to be considered during partition rebalancing. Each partition needs an open server file, these are stored in memory so it is not infinite. Lastly, replication latency for the cluster is increased.

>When you say the topic isn't used often, do you mean for a specific tenant or generally?

I meant for a specific tenant.

>I'm not sure I understand how frequency of use is a factor in the performance concerns you refer to.

I bring up the frequency because the some of the overhead I described above will be incurred whether a topic is used or not.

> This might be a system configuration / design rabbit hole. If a tenant does not do imports, it should be able to avoid having the topics specific to data import. That it cannot, might be an trade off of a separate design decision.

The main point I am conveying is to have better utilization of the resources available. Yes, rabbit holes are frowned upon!

>It might be more efficient from a storage perspective in Kafka. A trade off with that is that it potentially allows activity in one tenant to degrade the experience in another tenant, e.g. a large import in one tenant could hold up other imports in other tenants.

The current structure does not solve for one tenant holding up other tenants, since all tenant topics are usually serviced by one module instead of tenant specific module instances. There are strategies to mitigate bottlenecks in tenant interactions when using a single topic for all tenants.

Reply:
> Each partition needs an open server file, these are stored in memory so it is not infinite. 

Thanks, that's helpful.

> The main point I am conveying is to have better utilization of the resources available.

I'm fully supportive of the goal.

> Yes, rabbit holes are frowned upon!

I tend to want to go both broad and deep before returning to the topic at hand, as I find it important that decisions are framed in context. I'm aware that is an unusual approach and some folks find it frustrating.

> The current structure does not solve for one tenant holding up other tenants, since all tenant topics are usually serviced by one module instead of tenant specific module instances.

That depends upon how the consuming module is implemented and whether consumption of messages for different tenants are concurrent. If there are separate consumers for each tenant, then they might not be held up. That approach isn't an option (at least not in the same way) with this proposal.

> There are strategies to mitigate bottlenecks in tenant interactions when using a single topic for all tenants.

What would an example strategy?

Reply:
Maybe this is another topic that should go on the further discussion list and be discussed at a later stage of the proposal.

Reply:
>What would an example strategy?

A strategy can involve using a partition key to direct messages to specific partition. A partition key like instance id would distribute instances from every tenant so one tenant would not end up being processed exclusively. Increasing the partition count after setting the partition key will distribute the load further. A more involved strategy involves assigning priorities to partitions which producers and consumers are aware of. A set of partitions would be "fast lane" while others would be "regular".

I have added tenant throttling as an open question in the doc.